pykafka/producer.py

```diff
@@ -111,15 +158,112 @@ def __repr__(self):
             id_=hex(id(self))
         )
 
-    def _send_request(self, broker, req, attempt):
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish. If __exit__ is being
+        called because the with: block has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        self.stop(wait=self._worker_exception is None)
```

Comment:
You know, this is what happens when I don't read the decorator closely.  :)

I'm thought it did stuff before the function call, not after, so that wasn't necessarily true. Of course, the decorator is gone now, so the point is moot.